src/distinct.rs

```diff
@@ -227,13 +182,104 @@ pub fn rearrange_sequence(colors: &mut Vec<Color>, metric: DistanceMetric) {
         DistanceMetric::CIEDE2000 => c1.distance_delta_e_ciede2000(c2),
     };
 
+    // vector where the i-th element contains the minimum distance to the colors from 0 to i-1.
+    let mut min_distances = vec![i32::max_value(); colors.len()];
+
     for i in 1..colors.len() {
-        let (left, right) = colors.split_at_mut(i);
-        right.sort_unstable_by_key(|c| {
-            left.iter()
-                .map(|c2| (-distance(c2, c) * 1000.0) as i32)
-                .max()
-        });
+        let mut max_i = colors.len();
+        let mut max_d = i32::min_value();
+
+        for j in i..colors.len() {
+            min_distances[j] =
+                min_distances[j].min((distance(&colors[j], &colors[i - 1]) * 1000.0) as i32);
+
+            if min_distances[j] > max_d {
+                max_i = j;
+                max_d = min_distances[j];
+            }
+        }
+
+        colors.swap(i, max_i);
+        min_distances.swap(i, max_i);
+    }
+}
+
+impl DistanceResult {
+    fn new(colors: &[(Color, Lab)], distance_metric: DistanceMetric) -> Self {
+        let mut result = DistanceResult {
+            closest_distances: vec![(scalar::MAX, std::usize::MAX); colors.len()],
+            closest_pair: (std::usize::MAX, std::usize::MAX),
+            mean_closest_distance: 0.0,
+            min_closest_distance: scalar::MAX,
+            distance_metric,
+        };
+
+        for i in 0..colors.len() {
+            result.update_distances(colors, i, false);
+        }
+        result.update_totals();
+
+        result
+    }
+
+    fn update(&self, colors: &[(Color, Lab)], changed_color: usize) -> Self {
+        let mut result = self.clone();
+        result.update_distances(colors, changed_color, true);
+        result.update_totals();
+        result
+    }
+
+    fn update_distances(&mut self, colors: &[(Color, Lab)], color: usize, changed: bool) {
```

Comment:
minor: maybe rename `color` to `updated_color` or `modified_color`? Also, we could maybe rename `changed` to `perform_recursive_update`?

Reply:
Initially I named something like that, but `update_distances` is really just a way to recalculate the distances and it can be called with colors that aren't actually changed (like in the case of `to_recalc`) so that's why I generally named the parameter `color`. Wrt `changed` instead I'm ok with calling it `perform_recursive_update` even though I think I still prefer `changed` because I think the caller of the functions is more likely to understand what `changed` means than `perform_recursive_update` because it would have to understand that `perform_recursive_update` needs to be done whenever the color has changed.

Anyway, if you prefer the new names I'm ok with changing them.

Reply:
No, that makes sense. Let's go with your naming :+1: